### PR TITLE
instrumented Gateway

### PR DESCRIPTION
**Pull request by LoekL**
This PR instruments the Gateway, and includes some changes to the data model so that `partyJoinRequests` contain a `matchRequestId`, and a playerJoinRequest contains a `partyId` and a `matchRequestId` - which facilitates better tracking.

Hook overview: https://docs.google.com/spreadsheets/d/1nlEWy5b97IMdpmHXWLrv5ZtuU3YnU3Ba6-JmVBxB238/edit#gid=0 (where `eventSource` = `gateway_gateway`).

### CODE REVIEW

File: services/csharp/Gateway/OperationsServiceImpl.cs [hunk 1]
```diff
@@ -143,6 +146,33 @@ public override async Task<Empty> DeleteOperation(DeleteOperationRequest request
...
 
                     Reporter.CancelOperationInc();
+                    
+                    var eventAttributes = new Dictionary<string, string>
+                    {
+                        { "partyId", partyJoinRequest.Id },
+                        { "matchRequestId", partyJoinRequest.MatchRequestId },
+                        { "queueType", partyJoinRequest.Type }
+                    };
+                    string[] eventTypes = { "party_match_request_cancelled", "player_cancels_match_request" };
```
Comment (jared-improbable) [inline services/csharp/Gateway/OperationsServiceImpl.cs]: Can use `var` here and below

File: services/csharp/Gateway/OperationsServiceImpl.cs [hunk 2]
```diff
@@ -143,6 +146,33 @@ public override async Task<Empty> DeleteOperation(DeleteOperationRequest request
...
+                    foreach (string eventType in eventTypes)
```
Comment (jared-improbable) [inline services/csharp/Gateway/OperationsServiceImpl.cs]: It feels odd to have a `foreach` here for a two-item list, one of which is already special-cased...Why not just unroll this for now?

File: services/csharp/Gateway/OperationsServiceImpl.cs [hunk 3]
```diff
@@ -105,7 +108,7 @@ public override async Task<Empty> DeleteOperation(DeleteOperationRequest request
             if (!string.Equals(request.Name, playerIdentity))
             {
                 throw new RpcException(new Status(StatusCode.PermissionDenied,
-                    "Deleting another player's operation is forbidden."));
+                    $"Deleting another player's operation is forbidden: {request.Name} vs. {playerIdentity}"));
```
Comment (jared-improbable) [inline services/csharp/Gateway/OperationsServiceImpl.cs]: Drive-by? Does this need to stay in?

Comment (LoekL) [reply services/csharp/Gateway/OperationsServiceImpl.cs]: Was to help debug some things, but I'll revert back to original. Will leave upgrading our error messages for a potential future PR that takes everything into consideration together.

File: services/csharp/Gateway/OperationsServiceImpl.cs (hunk 1 above)
Comment (LoekL) [reply services/csharp/Gateway/OperationsServiceImpl.cs]: Like so:

```
var eventTypes = new[] { "party_match_request_cancelled", "player_cancels_match_request" };
foreach (var eventType in eventTypes)
```

Why is this better?

File: services/csharp/Gateway/OperationsServiceImpl.cs (hunk 2 above)
Comment (LoekL) [reply services/csharp/Gateway/OperationsServiceImpl.cs]: It would become: 

```
_analytics.Send("player_cancels_match_request", eventAttributes, partyJoinRequest.Party.LeaderPlayerId);
eventAttributes.Add("partyPhase", "Forming");
_analytics.Send("party_match_request_cancelled", eventAttributes, partyJoinRequest.Party.LeaderPlayerId);
```

I do not have a strong opinion on it. The foreach loop makes it very clear you're only adding `("partyPhase", "Forming")` for the party event, but it's 9 lines (vs. 3)..

Comment (jared-improbable) [reply services/csharp/Gateway/OperationsServiceImpl.cs]: Using a `foreach` like is akin to building a mini state machine, and the code just doesn't warrant that.

I'd also send a copy of `eventAttributes`, or there's a world where we add a third event type, and accidentally send along unrelated data
`new Dictionary<string,string>(eventAttributes) { "partyPhase", "Forming" }`
That also makes it clearer you're adding some extra data

Comment (LoekL) [reply services/csharp/Gateway/OperationsServiceImpl.cs]: Updated to:

```
_analytics.Send("player_cancels_match_request", eventAttributes, partyJoinRequest.Party.LeaderPlayerId);
var eventAttributesParty = new Dictionary<string,string>(eventAttributes) { {"partyPhase", "Forming"} };
_analytics.Send("party_match_request_cancelled", eventAttributesParty, partyJoinRequest.Party.LeaderPlayerId);
```

File: services/csharp/Gateway/OperationsServiceImpl.cs (hunk 1 above)
Comment (LoekL) [reply services/csharp/Gateway/OperationsServiceImpl.cs]: ^ No longer relevant.
